### PR TITLE
feat: implement autoscaling and the leastConnections strategy

### PR DESCRIPTION
**Pull request by congoamz**
### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 1]
```diff
@@ -21,39 +21,71 @@
...
+  private final IsItemValidFunc<V> isItemValidFunc;
+  private final ItemDisposalFunc<V> itemDisposalFunc;
 
   public CacheMap() {
+    this.isItemValidFunc = null;
+    this.itemDisposalFunc = null;
+  }
+
+  public CacheMap(
+      final IsItemValidFunc<V> isItemValidFunc,
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:40]: canDisposeFunc ?

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 2]
```diff
@@ -76,29 +108,52 @@ public int size() {
...
+    private long expirationTime;
 
     public CacheItem(final V item, final long expirationTime) {
       this.item = item;
       this.expirationTime = expirationTime;
     }
 
     boolean isExpired() {
+      if (isItemValidFunc != null) {
+        return System.nanoTime() > expirationTime && !isItemValidFunc.isValid(this.item);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:147]: Hmm I'm not sure what's the meaning of isItemValidFunc. Need to discuss.

File: wrapper/src/test/java/integration/refactored/TestInstanceInfo.java [hunk 3]
```diff
@@ -23,32 +23,36 @@
 public class TestInstanceInfo {
 
   private String instanceId; // "instance-1"
-  private String endpoint; // "instance-1.ABC.cluster-XYZ.us-west-2.rds.amazonaws.com"
-  private int endpointPort;
+  private String host; // "instance-1.ABC.cluster-XYZ.us-west-2.rds.amazonaws.com"
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/TestInstanceInfo.java:26]: This change is probably unrelated to leastConnections strategy. I'd make it in a separate PR if needed. We have quite stable integration suite and renaming of properties is definitely can do but it should be a strong reason for doing that.

File: wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        newInstanceConn.setReadOnly(true);
+        final String readerId = auroraUtil.queryInstanceId(newInstanceConn);
+
+        assertEquals(newInstance.getInstanceId(), readerId);
+        // Verify that there is a pool for the new instance
+        assertTrue(provider.getHosts().stream()
+            .anyMatch((url) -> url.equals(newInstance.getUrl())));
+        newInstanceConn.setReadOnly(false);
+      } finally {
+        auroraUtil.deleteInstance(newInstance);
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java:149]: What's the consequences for other integrations tests?

File: wrapper/src/test/java/integration/refactored/container/tests/ReadWriteSplittingTests.java [hunk 5]
```diff
@@ -600,11 +592,14 @@ public void test_pooledConnection_reuseCachedConnection() throws SQLException {
...
-    final HikariConfig config = new HikariConfig();
-    config.setMaximumPoolSize(1);
-    config.setInitializationFailTimeout(20000);
-    return config;
+  protected HikariPoolConfigurator getHikariConfig(int maxPoolSize) {
+    return (hostSpec, props) -> {
+      final HikariConfig config = new HikariConfig();
+      config.setMaximumPoolSize(maxPoolSize);
+      config.setInitializationFailTimeout(75000);
+      config.setKeepaliveTime(30000);
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/refactored/container/tests/ReadWriteSplittingTests.java:600]: TCPKeepAlive feature doesn't work well with ToxyProxy. Is it essential setting for the tests?

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 6]
```diff
@@ -21,51 +21,106 @@
...
   }
 
   public V get(final K key) {
-    final CacheItem<V> cacheItem = cache.computeIfPresent(key, (kk, vv) -> vv.isExpired() ? null : vv);
+    removeIfExpired(key);
+    final CacheItem cacheItem = cache.get(key);
     return cacheItem == null ? null : cacheItem.item;
   }
 
   public V get(final K key, final V defaultItemValue, final long itemExpirationNano) {
```
Comment (congoamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:52]: do we need both this and computeIfAbsent? They are quite similar. Differences: computeIfAbsent calls cleanup and extends the expiration

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 7]
```diff
@@ -21,51 +21,106 @@
...
+    removeIfExpired(key);
+    CacheItem cacheItem = cache.get(key);
+    if (cacheItem == null) {
+      cacheItem = new CacheItem(defaultItemValue, System.nanoTime() + itemExpirationNano);
+      cache.put(key, cacheItem);
+    }
     return cacheItem.item;
   }
 
+  public V getWithExtendExpiration(final K key, final long itemExpirationNano) {
```
Comment (congoamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:62]: only used in tests, should we remove?

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 8]
```diff
@@ -21,51 +21,106 @@
...
+  }
+
   public void put(final K key, final V item, final long itemExpirationNano) {
-    cache.put(key, new CacheItem<>(item, System.nanoTime() + itemExpirationNano));
     cleanUp();
+    removeAndDispose(key);
+    cache.put(key, new CacheItem(item, System.nanoTime() + itemExpirationNano));
   }
 
   public void putIfAbsent(final K key, final V item, final long itemExpirationNano) {
```
Comment (congoamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:88]: only used in tests, should we remove?

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 9]
```diff
@@ -21,51 +21,106 @@
...
+  }
+
+  public CacheMap(
+      final IsItemValidFunc<V> isItemValidFunc,
+      final ItemDisposalFunc<V> itemDisposalFunc) {
+    this.isItemValidFunc = isItemValidFunc;
+    this.itemDisposalFunc = itemDisposalFunc;
   }
 
   public V get(final K key) {
```
Comment (congoamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:46]: Should we call cleanup() in the get functions or any other function? Why do we call it only in certain functions?

File: wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java [hunk 10]
```diff
@@ -21,51 +21,106 @@
...
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
 
 public class CacheMap<K, V> {
 
-  private final Map<K, CacheItem<V>> cache = new ConcurrentHashMap<>();
-  private final long cleanupIntervalNanos = TimeUnit.MINUTES.toNanos(10);
+  private final Map<K, CacheItem> cache = new ConcurrentHashMap<>();
+  private long cleanupIntervalNanos = TimeUnit.MINUTES.toNanos(10);
```
Comment (congoamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/CacheMap.java:29]: Should this be configurable via the HikariPooledConnectionProvider constructor? Right now there isn't a way for the user to set this

File: wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+   * cleanup time or {@link ShouldDisposeFunc} indicated the entry should not be closed, the entry
+   * will be marked as non-expired.
+   *
+   * @param key                the key with which the specified value is to be associated
+   * @param mappingFunction    the function to compute a value
+   * @param itemExpirationNano the expiration time of the new or renewed entry
+   * @return the current (existing or computed) value associated with the specified key, or null if
+   *     the computed value is null
+   */
+  public V computeIfAbsent(
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/util/SlidingExpirationCache.java:72]: Should we remove similar method (and sliding expiration related functionality) from CacheMap<> ?

File: docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md [hunk 12]
```diff
@@ -39,7 +39,9 @@ The wrapper driver currently uses [Hikari](https://github.com/brettwooldridge/Hi
...
+You can optionally pass in a `HikariPoolMapping` function as a second parameter to the `HikariPooledConnectionProvider`. This allows you to decide when new connection pools should be created by defining what is included in the pool map key. A new pool will be created each time a new connection is requested with a unique key. By default, a new pool will be created for each unique instance-user combination. If you would like to define a different key system, you should pass in a `HikariPoolMapping` function defining this logic. Please see [ReadWriteSplittingPostgresExample.java](../../../examples/AWSDriverExample/src/main/java/software/amazon/ReadWriteSplittingPostgresExample.java) for an example of how to configure the pool map key.
```
Comment (karenc-bq) [inline docs/using-the-jdbc-driver/using-plugins/UsingTheReadWriteSplittingPlugin.md:42]: Can we provide a simple example here for the hikaripoolmapping function, and say, for the full example please see ReadWriteSplittingPostgresExample.java
```java
   props.setProperty("somePropertyValue", "1"); // used in getPoolKey
     final HikariPooledConnectionProvider connProvider =
         new HikariPooledConnectionProvider(
             ReadWriteSplittingPostgresExample::getHikariConfig,
             ReadWriteSplittingPostgresExample::getPoolKey
         );
    ConnectionProviderManager.setConnectionProvider(connProvider);

  private static String getPoolKey(HostSpec hostSpec, Properties props) {
    // Include the URL, user, and somePropertyValue in the connection pool key so that a new
    // connection pool will be opened for each different instance-user-somePropertyValue
    // combination.
    final String user = props.getProperty(PropertyDefinition.USER.name);
    final String somePropertyValue = props.getProperty("somePropertyValue");
    return hostSpec.getUrl() + user + somePropertyValue;
  }
```

File: wrapper/src/main/resources/messages.properties [hunk 13]
```diff
@@ -189,11 +197,10 @@ PropertyUtils.setMethodDoesNotExistOnTarget=Set method for property ''{0}'' does
 PropertyUtils.failedToSetProperty=Failed to set property ''{0}'' on target ''{1}''.
 PropertyUtils.failedToSetPropertyWithReason=Failed to set property ''{0}'' on target ''{1}''. {2}
 
-# RandomHostSelector
-RandomHostSelector.noHostsMatchingRole=No hosts were found matching the requested ''{0}'' role.
-
 # Read Write Splitting Plugin
 ReadWriteSplittingPlugin.setReadOnlyOnClosedConnection=setReadOnly cannot be called on a closed connection.
+ReadWriteSplittingPlugin.errorSwitchingToCachedReader=An error occurred while trying to switch to a cached reader connection to ''{0}''. The driver will attempt to establish a new reader connection.
+ReadWriteSplittingPlugin.errorSwitchingToCachedReaderWithCause=An error occurred while trying to switch to a cached reader connection to ''{0}''. Error message: ''{1}''. The driver will attempt to establish a new reader connection.
```
Comment (karenc-bq) [inline wrapper/src/main/resources/messages.properties:203]:
```suggestion
ReadWriteSplittingPlugin.errorSwitchingToCachedReaderWithCause=An error occurred while trying to switch a cached reader connection to ''{0}''. Error message: ''{1}''. The driver will attempt to establish a new reader connection.
```

File: wrapper/src/main/resources/messages.properties [hunk 14]
```diff
@@ -189,11 +197,10 @@ PropertyUtils.setMethodDoesNotExistOnTarget=Set method for property ''{0}'' does
 PropertyUtils.failedToSetProperty=Failed to set property ''{0}'' on target ''{1}''.
 PropertyUtils.failedToSetPropertyWithReason=Failed to set property ''{0}'' on target ''{1}''. {2}
 
-# RandomHostSelector
-RandomHostSelector.noHostsMatchingRole=No hosts were found matching the requested ''{0}'' role.
-
 # Read Write Splitting Plugin
 ReadWriteSplittingPlugin.setReadOnlyOnClosedConnection=setReadOnly cannot be called on a closed connection.
+ReadWriteSplittingPlugin.errorSwitchingToCachedReader=An error occurred while trying to switch to a cached reader connection to ''{0}''. The driver will attempt to establish a new reader connection.
```
Comment (karenc-bq) [inline wrapper/src/main/resources/messages.properties:202]:
```suggestion
ReadWriteSplittingPlugin.errorSwitchingToCachedReader=An error occurred while trying to switch a cached reader connection to ''{0}''. The driver will attempt to establish a new reader connection.
```

File: wrapper/src/test/java/integration/refactored/container/ConnectionStringHelper.java [hunk 15]
```diff
@@ -84,16 +85,25 @@ public static String getWrapperUrl() {
...
   }
 
+  public static String getWrapperUrl(TestInstanceInfo instance) {
+    return getWrapperUrl(
+        TestEnvironment.getCurrent().getCurrentDriver(),
+        instance.getHost(),
+        instance.getPort(),
+        TestEnvironment.getCurrent().getInfo().getDatabaseInfo().getDefaultDbName());
+  }
+
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/refactored/container/ConnectionStringHelper.java:105]:
```suggestion
```

File: wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+  protected static Properties getProps() {
+    final Properties props = getDefaultPropsNoPlugins();
+    PropertyDefinition.PLUGINS.set(props, "auroraHostList,readWriteSplitting");
+    return props;
+  }
+
+  protected static Properties getPropsWithFailover() {
+    final Properties props = getDefaultPropsNoPlugins();
+    PropertyDefinition.PLUGINS.set(props, "readWriteSplitting,failover,efm");
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java:93]: Do you need to set monitoring-socketTimeouts here?

File: wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+    ConnectionProviderManager.setConnectionProvider(provider);
+
+    final List<Connection> connections = new ArrayList<>();
+    try {
+      for (int i = 1; i < instances.size(); i++) {
+        final String connString = ConnectionStringHelper.getWrapperUrl(instances.get(i));
+        final Connection conn1 = DriverManager.getConnection(connString, props);
+        connections.add(conn1);
+        final Connection conn2 = DriverManager.getConnection(connString, props);
+        connections.add(conn2);
```
Comment (karenc-bq) [inline wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java:209]:
```suggestion
        // Create 2 connections per instance.
        connections.add(DriverManager.getConnection(connString, props));
        connections.add(DriverManager.getConnection(connString, props));
```

File: wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java (hunk 16 above)
Comment (congoamz) [reply wrapper/src/test/java/integration/refactored/container/tests/AutoscalingTests.java:93]: Don't think so because we are setting the regular socket timeout in getDefaultPropsNoPlugins

File: wrapper/src/main/resources/messages.properties (hunk 13 above)
Comment (congoamz) [reply wrapper/src/main/resources/messages.properties:203]: I think the error message is misleading if we remove 'to'. Removing it makes it sound like the error occurred when we tried to change a cached reader connection to another reader. The actual scenario is that we tried to switch from a writer to a cached reader. I'll rephrase the error message a bit to hopefully be clearer

**Review by karenc-bq — APPROVED** [review]
